Documentation/design-docs/host-tracing.md

```diff
@@ -0,0 +1,74 @@
+
+Currently (as of .NET Core 2.1) all these components include the same code for tracing but act completely independently. VIa the fact that they share the same code, the end behavior is that they all write their trace output to the same location `stderr`. The only synchronization currently implemented is that each component fully flushes all internal tracing buffers before transferring control to a different component.
+
+## New trace routing
+By default tracing is disabled and is not routed anywhere. If it's enabled it is routed through an instance of `host_trace_listener` interface.
+``` C++
+struct host_trace_listener
+{
+    void trace_verbose(const pal::char_t *message);
+    void trace_info(const pal::char_t *message);
+    void trace_warning(const pal::char_t *message);
+    void trace_error(const pal::char_t *message);
+    void flush();
+}
+```
+
+The `message` parameter is a standard `NUL` terminated string.
+
+The host components provide two built in implementations of this interface:
+* `stderr_trace_listener` which implements the .NET Core 2.1 behavior of routing tracing to the `stderr` output of the process. This is the default listener which is activated by setting `COREHOST_TRACE=1` environment variable.
+* `file_trace_listener` which routes tracing into a file. This listener is activated by setting the `COREHOST_TRACE=1` and also providing the full path to the trace file in `COREHOST_TRACEFILE=<path>` environment variable.
+
+Custom host can provide its own implementation of `host_trace_listener` by calling:
+``` C++
+void hostfxr_set_trace_listener(host_trace_listener* listener)
+```
+from the `hostfxr` component or
+``` C++
+void corehost_set_trace_listener(host_trace_listener* listener)
+```
+from the `hostpolicy` component.
+
+The functions behave exactly the same in both components. The `listener` parameter can be:
+* a pointer to an implementation of `host_trace_listener` which is then registered the only listener for all tracing.
+* `NULL` value which unregisters any previously registered listener. After this call tracing is disabled.
+
+Custom host can and should register the trace listener as the first thing it does with the respective host component to ensure that all tracing is routed to it.  
+
+Only one trace listener can be registered at any given time.  
+
+Registering custom trace listener or setting it to `NULL` overrides any tracing enabled by environment variables.
+
+The `hostfxr` component will propagate the trace listener to the `hostpolicy` component before it calls into it. So custom host only needs to register its trace listener with the `hostfxr` component and not both. The propagation of the trace listener is only done for the duration necessary after which it will be unregistered again. So custom host might need to register its own listener if it makes calls directly to `hostpolicy` on top of the calls to `hostfxr`.
```

Comment:
Is it possible to have a design where there is one place that controls registration/knobs of callbacks? hostpolicy/fxr can still implement their own triggers. This will simplify the story for app/host developers.

Reply:
@swaroop-sridhar I'm not sure what you mean. This paragraph specifically tries to address the issue with potential simultaneous usage of `hostfxr` and `hostpolicy` exports directly from the same custom host. I think that this would be pretty rare, as custom host is likely to chose only one level of abstraction (either `hostfxr` or `hostpolicy`), but unlikely both. Alternatively custom host can call methods from both, but it would typically not use those where `hostfxr` ends up calling `hostpolicy`. So in general I view this as a corner case scenario.

As it stands right now, I'm not sure it's possible to have one place where the listener registration would happen for both components. Mostly because there are scenarios where only one component is loaded and the other simply doesn't exist. And there's no shared slot which can be accessed by both components.

Reply:
Thanks @vitek-karas; My question was whether it is worh having a shared component, in order to facilitate common registration. Having this component is probably not worth just for tracing, but if more shared infrastructure is required in future, we can have it.

Reply:
Having a shared component between `hostfxr` and `hostpolicy` would be rather tricky. They live in different version bubbles. (I'm simplifying, but still) `hostfxr` is "global" on the machine and the latest version always. `hostpolicy` is per runtime. Given this I actually think we should try to avoid shared component as much as possible.